rc/lsp.kak

```diff
@@ -230,18 +230,29 @@ define-command lsp-hover -docstring "Request hover info for the main cursor posi
     lsp-did-change-and-then lsp-hover-request
 }
 
-define-command -hidden lsp-hover-request -docstring "Request hover info for the main cursor position" %{
-    nop %sh{ (printf '
+define-command lsp-hover-in-hover-client -docstring "Request hover info for the main cursor position in a separate client" %{
```

Comment:
The name repeats "hover", which isn't so nice.
Also "client" should be something like "buffer" (as opposed to the info box), also because it's a valid use case to have the `*hover*` buffer appear in the current client.

That's another thing that we might want to tweak:
by hardcoding the client name, it seems difficult to use the current client.
Granted, we can use `:rename-client hoverclient`, but that will fail if some other client is already called `hoverclient` from a previous call.
This makes the `docsclient` approach seem more attractive, because then we can use `set global docsclient %val{client}; lsp-hover-in-hover-client`

Anyway, if we don't change the interface, I'd say

```suggestion
define-command lsp-hover-buffer -docstring "Show hover info for the main cursor position in a scratch buffer" %{
```

Reply:
> Granted, we can use :rename-client hoverclient, but that will fail if some other client is already called hoverclient from a previous call.

Is that possible? The renaming of the client is only happening in `catch` statement. And we landed up in the catch only because the `hoverclient` did not exist...

Reply:
I mean that the user can rename any client to `hoverclient` if they want the `*hover*` buffer to show up there, but an option would be more flexible

Reply:
Regarding the other discussion about `%opt{docsclient}` - I think we should just accept an optional argument, and in the default mapping call `:lsp-hover-buffer %opt{docsclient}`.
Then default to `hoverclient` if docsclient is not set.

(Of course I can do this extension - I just like to get the commits into a "mostly done" state before merging so the repo history is easy to read)

Reply:
@krobelus 

> (Of course I can do this extension - I just like to get the commits into a "mostly done" state before merging so the repo history is easy to read)

I should have time in the next few days to deal with the review comments you have made. Thanks again for your useful and thoughtful review comments.

- I'll push up my changes in a very small number of commits on this branch, force pushing if necessary on this branch
- You are free to make further adjustments/improvements as _separate_ commits on the same branch (you should be able to push to this branch)
- We can then merge this branch when you're happy

Personally I would prefer that you don't `--amend` my commits to make any adjustments. Amending does keep the history simple if but confuses me because I may see a change and then not be sure if that is something you added in or I added and then I have to refer back to the original git history.

The consolidated feature code can always be viewed in the merge commit so we always have the ability to see what the code for the whole feature is.

Does this sound reasonably? We can also discuss the git workflow on email if you want...

Reply:
P.S. And the merge from this branch should ideally be a normal merge rather than a squash merge -- that keeps things clear about what changes I made and what changes you made -- again we can discuss on email if needed...

Reply:
> Personally I would prefer that you don't `--amend` my commits to make any adjustments. Amending does keep the history simple if but confuses me because I may see a change and then not be sure if that is something you added in or I added and then I have to refer back to the original git history.

Cleaning up history may be a slight inconvenience to the author and first reviewer, but it helps *everyone* else, including future-you and future-me.
People who read the history later will arguably prefer a condensed history, and probably don't care about our old mistakes.

There are tools like git-range-diff that make it easier to see how history was rewritten.

I can read the history of a repository like https://github.com/git/git and quickly know what is actually going on because there is very little noise in the history (because they use a proper review process).
Of course we are much much smaller, so a clean history is not essential but it's still much nicer.

I must admit that I couldn't wait to merge the next/prev-symbol PR, so I ended up doing the cleanups myself.
Hence the confusion -- sorry -- I should have waited for your changes, and communicated how we should structure them.
I'm happy to review until we have something we're both happy with.

However, if review changes are added in separate commits with not-so-useful commit messages, that just makes the final history worse than if I just did the cleanups myself. (Also for example, the back-and-forth changes to `lsp_symbol_kind_completion` should go away because they cancel each other out.)

If you strongly prefer never rewriting your commits, that's okay too, then I'll probably just apply most review comments myself later.
I just don't see a good reason to treat old commits from pending PRs as if they are set in stone. For example, if a question comes up in code review that should be answered by the commit message, why not change the message.

> The consolidated feature code can always be viewed in the merge commit so we always have the ability to see what the code for the whole feature is.

In theory yes, but in practice git-blame will still point to a single commit, that would be more useful if the commit was a logical unit.

> P.S. And the merge from this branch should ideally be a normal merge rather than a squash merge -- that keeps things clear about what changes I made and what changes you made

I'm still not sure if it's better to have merge commits or not, but they seem fine.

Reply:
The problem here is that changing existing commits always feels to me like pulling the rug from under your contributors.
If I've been working on review comments or new stuff for a PR for a day or two until I consider it ready for pushing, `! [rejected]` is not fun to see. Especially when accompanied by a bunch of diffs that require me to change my mental model of the code I've been working with.

A clean history on merge does make sense, but, in my opinion, it should either be discussed in comments until all parties are on the same page or be postponed until the PR author considers their work complete and ready for merge/hand-off.
I always start out my PRs with "Allow edits" unchecked for that reason.

Reply:
@sclu1034 @krobelus @ul
> The problem here is that changing existing commits always feels to me like pulling the rug from under your contributors.

I agree with you and I am happy that you have independently arrived at this conclusion without you and me knowing each other. 

It seems `--amend`ing commits of others and changing code and git commit comments seems to be a regular occurrence on this project.

In fact my singular commit on this branch has been again modified in "successor" #570. See how different https://github.com/kak-lsp/kak-lsp/pull/570/commits/a648be06bb6e78087d7ff57e621525fc8af997ff is to my original commit https://github.com/kak-lsp/kak-lsp/commit/15fabf5a2ec0ed1478ee245513b0684c60be1b37

This is even though I have made an explicit request in private communication that I would prefer that @krobelus is free to build upon my commit in _subsequent_ commits and make any changes that are deemed necessary. I just want my _original_ commit to be preserved.

The same story happened in https://github.com/kak-lsp/kak-lsp/pull/557 -- My commit was modified significantly and I have already objected to that in private email communication. At that point I was not aware this approach of modifying other people's code in this project was common and regular. Puzzled, thinking this was a "one-off" I even was happy to have Johannes as a Co-Author. See here https://github.com/kak-lsp/kak-lsp/pull/557#issuecomment-969280030 . Yet even that request was not accepted and the commit was finally merged as https://github.com/kak-lsp/kak-lsp/commit/369ce314ce4f1ff269f13c2616a4dfbcbeb733d0 with the commentary changed and comment "
Patch-tweaked-by: Johannes Altmanninger <aclopte@gmail.com>". I certainly don't think that Johannes, tweaked my patch. There were a significant amount of changes made.

This is a pattern that is being repeated. All this is being done in the name of "clean commit history". What use is "clean commit history" if your contributors are unhappy? I would assume that "clean commit history" would be secondary.

These are my requests:
- My commits are not `--amend` without discussing with me. Which means that in the default case, please leave them as is. I am happy for your to combine them together also or ask me to combine them at any point in time.
- **If the maintainer or anybody wants to improve or fix my code, they are free to do them on _separate commits_**. If I take too long in incorporating the review comments, they can discuss with me on next steps. Failing prompt resolution, I am happy for my commits to be used as a base to incorporate additional features, and fixes in _separate_ commits. My objection is _only_ about `--amend`ing my commit, changing my commit commentary and most importantly the code that I wrote.
- Avoid giving fiat-acomplis to contributors. Discuss what you want to do before you do it. If you want to merge, please go ahead and merge. But if you are merging with `--amend`-ing the commits, at least ask the original author first?
  
I don't think I'm asking very much. 

It is not my intention to slow down the pace of this project. All I'm saying is don't `--amend` my code without discussing. Do whatever you want on top of my commits in separate commits though. I'm cool with that.

Reply:
> A clean history on merge does make sense, but, in my opinion, it should either be discussed in comments until all parties are on the same page or be postponed until the PR author considers their work complete and ready for merge/hand-off.

Agreed. As I said earlier, I'm happy to discuss every change I'd like to make.

We can still do that for this PR (dropping #570), I'm not sure if @sidkshatriya is up for that?

Reply:
> Failing prompt resolution, I am happy for my commits to be used as a base to incorporate additional features, and fixes in _separate_ commits.

That doesn't feel right to me. If there is something wrong in a commit (and the author is no longer available) I'd rather fix the commit than creating a fixup commit.

Reply:
@krobelus I feel we have reached an impasse here. I am no longer motivated to spent my time and energy on this project because you refuse to simply use my commit but insist on on `--amend`ing it. 

I actually was intending to work on the review points but you took the initiative on #570 and incorporated the review comments yourself into the work I originally did. I comment you for taking the initiative. But to say that you refuse to build upon this commit unless you have the ability to `--amend` it is absolutely breathtaking to me.

**If there is something wrong in a commit, you have another commit to fix it. When a merge happens, your fixes will be merged along with it when the branch is merged. You are aware of this.**

This commit compiles, is functional. The feature works. Yes, it can do with improvement and changes in approach. It has flaws but I'm not aware of any fatal ones. But to say that it must not exist as _one_ of the many commits on a feature branch that you have full ability to add on before a merge is quite surprising to me. 

The experience has been a net negative for me. At this point I can continue debating with you or cut my losses. I suggest you do what you feel like without my approval or consent.

Or you can take a small pause. Don't take my word work it. Please consult with other contributors  (e.g. @ul  -- the originator of the project or @sclu1034 who happened to add an opinion) and see what they suggest. You can still chose to take your own path in the end!

Reply:
P.S. @krobelus In the interest of resolving this issue (because I just don't have the time to spend more time on a contribution I purely did out of love of a project)...

**Proposal**
Please make the minimum changes you deem acceptable to this commit to be considered merge-able. You have already done the work in #570 where you have addressed many of your own review comments. 

You can even commit as the primary author of this commit. I'm happy to be credited as the Co-author of this feature in the commit message. And then we can call it a day. kak-lsp becomes better and we can move on.

This would address your concern about having a commit on the master branch which does not have any flaws. Can we do this?

Reply:
Thanks for the contribution ;)